### PR TITLE
Grant staking policy

### PR DESCRIPTION
**Pull request by eth-r**
Replaces #1514 

Refs: #1478

Grants can now specify an arbitrary policy for calculating how much the grantee may stake.

### CODE REVIEW

**Comment by pdyraga** [comment]
There are three tests failing here.

**Comment by pdyraga** [comment]
Just two comments and we need to fix failing tests. Then I am happy to merge, the solution is very elegant. 

@r-czajkowski we'll need your help with updating KEEP token dashboard. There are two staking policies deployed as contracts:
- `GuaranteedMinimumStakingPolicy` 
- `PermissiveStakingPolicy`

I think it's fine to assume that by default, the first one is applied to all revocable grants and the second one to all non-revocable grants. Grant policy address is passed in `_extraData` in `receiveApproval` of `TokenGrant`.

This should affect only the section for the token owner for creating grants.

**Comment by r-czajkowski** [comment]
@pdyraga ready #1546

File: solidity/contracts/TokenGrant.sol [hunk 1]
```diff
@@ -230,8 +234,22 @@ contract TokenGrant {
             _revocable = true;
         }
 
+        address _stakingPolicy = _extraData.toAddress(117);
```
Comment (r-czajkowski) [inline solidity/contracts/TokenGrant.sol:237]: Let's add `_stakingPolicy` to docs.

File: solidity/contracts/GrantStakingPolicy.sol [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+/// @dev A staking policy defines the function `getStakeableAmount`
+/// which calculates how many tokens may be staked from a token grant.
+contract GrantStakingPolicy {
+    function getStakeableAmount(
+        uint256 _now,
+        uint256 grantedAmount,
+        uint256 duration,
+        uint256 start,
+        uint256 cliff,
+        uint256 withdrawn) public view returns (uint256);
```
[the remark this answers is not in the document]
Comment (eth-r) [reply solidity/contracts/GrantStakingPolicy.sol:13]: In my experience linter complains if I do move it to the next line. I haven't yet found another way to make the linter accept interface declarations.

File: solidity/test/token_grant/TestStakingPolicy.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  let stakingContract;
+  let minimumStake;
+  let largeGrant;
+  let mediumGrant;
+  let smallGrant;
+  let start = 1000;
+  let duration = 2000;
+  let cliff = 1500;
+
+  function units(n) { return minimumStake.divn(2000).muln(n); }
```
[the remark this answers is not in the document]
Comment (eth-r) [reply solidity/test/token_grant/TestStakingPolicy.js:69]: That was a quick fix to not have to entirely recalculate all of the amounts. Previously the minimum stake was 2000 and other amounts were relative to that.

[the remark this answers is not in the document]
Comment (eth-r) [reply solidity/test/token_grant/TestStakingPolicy.js]: Fixed in latest

File: solidity/contracts/GrantStakingPolicy.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/GrantStakingPolicy.sol:13]: Alright, let's leave it as-is and we'll let the linter config decide later :)

File: solidity/test/token_grant/TestStakingPolicy.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/test/token_grant/TestStakingPolicy.js]: So `10000` is a precision? Can we add a comment to `tokens` functions explaining it?

**Comment by pdyraga** [comment]
👀

**Comment by pdyraga** [comment]
Good catch with `EvilStakingPolicy`!

**Pull request merged**